pkg/ebpf/generic_map.go

```diff
@@ -0,0 +1,327 @@
+// inferred from the types of K and V.
+// Important: if the map is a per-cpu map, V must be a slice type
+func NewGenericMap[K interface{}, V interface{}](spec *ebpf.MapSpec) (*GenericMap[K, V], error) {
+	// Automatic inference of sizes. We assume that K/V are simple types that
+	// can be instantiated with no arguments
+	var kval K
+	var vval V
+
+	err := validateValueTypeForMapType[V](spec.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	spec.KeySize = uint32(unsafe.Sizeof(kval))
+
+	if isPerCPU(spec.Type) {
+		spec.ValueSize = uint32(reflect.TypeOf(vval).Elem().Size())
+	} else {
+		spec.ValueSize = uint32(unsafe.Sizeof(vval))
+	}
+
+	m, err := ebpf.NewMap(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GenericMap[K, V]{
+		m: m,
+	}, nil
+}
+
+func validateValueTypeForMapType[V interface{}](t ebpf.MapType) error {
+	var vval V
+	if isPerCPU(t) && reflect.TypeOf(vval).Kind() != reflect.Slice {
+		return errors.New("per-cpu maps require a slice type for the value")
+	}
+	return nil
+}
+
+// Map creates a new GenericMap from an existing ebpf.Map
+func Map[K interface{}, V interface{}](m *ebpf.Map) (*GenericMap[K, V], error) {
+	if err := validateValueTypeForMapType[V](m.Type()); err != nil {
+		return nil, err
+	}
+
+	return &GenericMap[K, V]{
+		m: m,
+	}, nil
+}
+
+// GetMap gets the generic map with the given name from the manager
+func GetMap[K interface{}, V interface{}](mgr *manager.Manager, name string) (*GenericMap[K, V], error) {
+	m, _, err := mgr.GetMap(name)
+	if err != nil {
+		return nil, err
+	}
+	if m == nil {
+		return nil, fmt.Errorf("map %s not found", name)
+	}
+	gm, err := Map[K, V](m)
+	if err != nil {
+		return nil, err
+	}
+	return gm, nil
+}
+
+// Map returns the underlying ebpf.Map
+func (g *GenericMap[K, V]) Map() *ebpf.Map {
+	return g.m
+}
+
+// IteratorOptions are options for the Iterate method
+type IteratorOptions struct {
+	BatchSize       int  // Number of items to fetch per batch. If 0, use default value (100)
+	ForceSingleItem bool // Force the use of the single item iterator even if the batch API is supported
+}
+
+// Put inserts a new key/value pair in the map. If the key already exists, the value is updated
+func (g *GenericMap[K, V]) Put(key *K, value *V) error {
+	if g.isPerCPU() {
+		return g.m.Put(unsafe.Pointer(key), *value)
+	}
+
+	return g.m.Put(unsafe.Pointer(key), unsafe.Pointer(value))
+}
+
+// Update updates the value of an existing key in the map.
+func (g *GenericMap[K, V]) Update(key *K, value *V, flags ebpf.MapUpdateFlags) error {
+	return g.m.Update(unsafe.Pointer(key), unsafe.Pointer(value), flags)
+}
+
+// Lookup looks up a key in the map and returns the value. If the key doesn't exist, it returns ErrKeyNotExist
+func (g *GenericMap[K, V]) Lookup(key *K, valueOut *V) error {
+	if g.isPerCPU() {
+		return g.m.Lookup(unsafe.Pointer(key), *valueOut)
+	}
+
+	return g.m.Lookup(unsafe.Pointer(key), unsafe.Pointer(valueOut))
+}
```

Comment:
shouldn't we have here LookupBatch and UpdateBatch?

I do see such methods for Delete, but not for the rest

Reply:
I added UpdateBatch, the lookup function is the one used for iteration (it's got a confusing name, BatchLookup doesn't really do the same thing as Lookup).